### DISCUSSION TITLE
HA Core Pull Request

### DISCUSSION THREAD

**Discussion by disforw** [kdeyev/eyeonwater#62] [General]
This post is a response to a question asked in PR #61 

That is a great question, but difficult to answer.
I have submitted a simple helper integration to HACS called "inverse" which has been waiting about 2 months in queue. There is no rhyme or reason to how he accepts new submissions. https://github.com/hacs/default/pull/1953

On the flip side, yes, Core is definitely more strict on code inspection, but their reviews will help us improve our integration and our coding. I submitted a PR to core for moving an integration from yaml to Config Flow and I needed 150 commits from reviews of moderators!  https://github.com/home-assistant/core/pull/80450

If we tackle it together, we can make it through and get some awesome help doing so...
HOWEVER, there are a couple of speed bumps to consider if we go to Core...

- You can only submit one platform when it's a new integration, so we'd have to choose sensor or binary sensor.
- we'd have to work through this statistics thing more elegantly
- we'd have to stick to a very strict code enforcement

**Comment by disforw** [comment]
I would love to say let's do BOTH, but the way the icons and images repository is structured, we need to only choose one. 😟

**Comment by kdeyev** [comment]
Thank you for sharing this information! 
I think we can try to get the integration into HA. I'm concerned a bit about the issue with negative values in statistics, but we are not in a rush so hopefully we will be able to tackle it properly. 
To say honestly I'm surprised by the limitation of one platform per integration, but can more platforms be added in later versions of the integration?
Regarding the code enforcement, I've added pre-commit checks, which should somehow help with it.

**Comment by kdeyev** [comment]
By the way, I'm working on adding unit tests to pyonwater, and eventually, it will be required to add tests to the HA integration code as well.

**Comment by disforw** [comment]
Of course we can add additional platforms later in a separate PR. It's just to make it easier to inspect... Trying to keep the PRs smaller for easier reviews.

**Comment by disforw** [comment]
As a matter of fact, for the initial PR, why don't we start with submitting just the meter read sensor without the statistics and add them in a later PR?

**Reply by disforw** [reply]
We can also add the binary Sensors in later as well

**Comment by kdeyev** [comment]
It makes a lot of sense, to drop everything but just plain sensor. The source code should be very straightforward and easy to review (and accept 😄)
But we may still need to add unit tests, I'm not sure if it's a requirement for the initial commit.

**Comment by disforw** [comment]
Would you like to open the PR to Core under yourself and add me as a collaborator? Or should I open and add you?

**Comment by kdeyev** [comment]
@disforw Please go ahead and open the PR.

**Comment by disforw** [comment]
I forked core and invited you to be a collaborator. We will create the PR from there. We need to recreate the integration under homeassistent/components/eyeonwater. 
I’m going to work on recreating the integration without the statistics or the service.
I’ll also attempt to submit a PR with both platforms and see if they ask us to remove one.

once we ppen a PR, any edits made in the fork by either of us should immediately hit the PR

**Comment by disforw** [comment]
One thing that just came up, we need a version of pyonwater where read_meter doesn’t require a “days_to_load” var

**Comment by kdeyev** [comment]
Yes, it's here: https://github.com/kdeyev/eyeonwater/pull/64
Meter_info and statistic are decoupled

**Comment by kdeyev** [comment]
```
    async def read_meters(self, days_to_load=3):
        """Read each meter."""
        for meter in self.meters:
            try:
                await meter.read_meter_info(client=self.client)
                await meter.read_historical_data(
                    client=self.client,
                    days_to_load=days_to_load,
                )
            except (EyeOnWaterAPIError, EyeOnWaterAuthError) as error:
                raise UpdateFailed(error) from error
        return self.meters

**Comment by disforw** [comment]
Ok, have a look at the fork i’ve invited you to. There are 9 commits which adds the files we need. You should be able to make commits which will pass right through to the PR when we open it.
I need to run pre-commit on it all and fix it up a bit.

**Comment by disforw** [comment]
**The PR has been opened!**
https://github.com/home-assistant/core/pull/99901
It immediately asked me to limit the initial PR to a single platform (as expected), so I did.
I already see some CLI failures, I’ll look at them tomorrow. You have editing rights via the shared repo (https://github.com/disforw/homeassistant-core), your commits should pass through if you want.

**Comment by kdeyev** [comment]
I'll take a look at the pre-commit errors

**Comment by disforw** [comment]
It seems to not like that I have two sensor classes in one sensor file, I'm going to remove the temperature sensor for this initial pull request. Later on I will read it and merge the sensor classes into one.

**Comment by kdeyev** [comment]
The pipeline is green

**Reply by kdeyev** [reply]
I also backported some changes: https://github.com/kdeyev/eyeonwater/pull/66
